File: apps/ios/AllyClock/Time/ClockFormatter.swift
import Foundation

enum ClockFormatter {
    typealias BigTime = (digits: String, ampm: String?)

    static func bigTime(_ date: Date, in timeZone: TimeZone, locale: Locale) -> BigTime {
        let isTwelveHour = usesTwelveHourClock(locale)

        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = timeZone
        formatter.setLocalizedDateFormatFromTemplate(isTwelveHour ? "hmm" : "Hmm")

        var rendered = formatter.string(from: date)
        if isTwelveHour {
            for symbol in [formatter.amSymbol, formatter.pmSymbol, "AM", "PM"].compactMap({ $0 }) {
                rendered = rendered.replacingOccurrences(of: symbol, with: "")
            }
            rendered = rendered.trimmingCharacters(in: .whitespaces)
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let hour24 = calendar.component(.hour, from: date)
        let amSymbol = nonEmpty(formatter.amSymbol, fallback: "AM")
        let pmSymbol = nonEmpty(formatter.pmSymbol, fallback: "PM")
        let ampm: String? = isTwelveHour ? (hour24 < 12 ? amSymbol : pmSymbol) : nil

        return (rendered, ampm)
    }

    static func precise(_ date: Date, in timeZone: TimeZone) -> String {
        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = timeZone
        let comps = calendar.dateComponents([.hour, .minute, .second, .nanosecond], from: date)
        let hundredths = min(99, Int((Double(comps.nanosecond ?? 0) / 10_000_000).rounded()))
        return String(
            format: "%02d:%02d:%02d.%02d",
            comps.hour ?? 0,
            comps.minute ?? 0,
            comps.second ?? 0,
            hundredths
        )
    }

    static func dateTZ(_ date: Date, in timeZone: TimeZone, locale: Locale) -> String {
        let formatter = DateFormatter()
        formatter.locale = locale
        formatter.timeZone = timeZone
        formatter.dateStyle = .long
        formatter.timeStyle = .none
        let datePart = formatter.string(from: date)
        return "\(datePart) · \(gmtOffset(timeZone, for: date))"
    }

    static func gmtOffset(_ timeZone: TimeZone, for date: Date) -> String {
        let totalSeconds = timeZone.secondsFromGMT(for: date)
        let sign = totalSeconds >= 0 ? "+" : "\u{2212}"
        let magnitude = abs(totalSeconds)
        let hours = magnitude / 3600
        let minutes = (magnitude % 3600) / 60
        return String(format: "GMT%@%02d:%02d", sign, hours, minutes)
    }

    private static func nonEmpty(_ value: String?, fallback: String) -> String {
        guard let value, !value.isEmpty else {
            return fallback
        }
        return value
    }

    private static func usesTwelveHourClock(_ locale: Locale) -> Bool {
        let normalized = locale.identifier.replacingOccurrences(of: "-", with: "_")
        return normalized.split(separator: "_").contains("US")
    }
}

File: apps/web/src/app/faces/fullscreen/clock-formatter.ts
export interface BigTime {
  digits: string;
  ampm: string | null;
}

export function bigTime(date: Date, locale: string, timeZone: string): BigTime {
  const parts = new Intl.DateTimeFormat(locale, {
    hour: 'numeric',
    minute: '2-digit',
    timeZone,
  }).formatToParts(date);

  const digits = parts
    .filter((part) => part.type === 'hour' || part.type === 'minute')
    .map((part) => part.value)
    .join(':');
  const ampm = parts.find((part) => part.type === 'dayPeriod')?.value ?? null;

  return { digits, ampm };
}

export function precise(date: Date, timeZone: string): string {
  const parts = new Intl.DateTimeFormat('en-US', {
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hourCycle: 'h23',
    timeZone,
  }).formatToParts(date);
  const get = (type: Intl.DateTimeFormatPartTypes): string =>
    parts.find((part) => part.type === type)?.value ?? '00';
  const hundredths = Math.min(99, Math.round(date.getMilliseconds() / 10));
  return `${get('hour')}:${get('minute')}:${get('second')}.${String(hundredths).padStart(2, '0')}`;
}

export function dateTZ(date: Date, locale: string, timeZone: string): string {
  const datePart = new Intl.DateTimeFormat(locale, { dateStyle: 'long', timeZone }).format(date);
  return `${datePart} · ${gmtOffset(date, timeZone)}`;
}

export function gmtOffset(date: Date, timeZone: string): string {
  const name =
    new Intl.DateTimeFormat('en-US', { timeZone, timeZoneName: 'longOffset' })
      .formatToParts(date)
      .find((part) => part.type === 'timeZoneName')?.value ?? 'GMT';
  const normalized = name === 'GMT' ? 'GMT+00:00' : name;
  return normalized.replace('-', '−');
}
